File: app/src/main/java/com/adrzdv/mtocp/ui/viewmodel/ViewModelFactoryProvider.java
package com.adrzdv.mtocp.ui.viewmodel;

import androidx.lifecycle.ViewModel;

import com.adrzdv.mtocp.App;
import com.adrzdv.mtocp.domain.repository.KriCoachRepo;

import java.util.HashMap;
import java.util.Map;

public class ViewModelFactoryProvider {

    public static CustomViewModelProvider provideFactory() {

        Map<Class<? extends androidx.lifecycle.ViewModel>,
                Provider<? extends ViewModel>> creators = new HashMap<>();

        creators.put(ViolationViewModel.class,
                () -> new ViolationViewModel(App.getViolationRepository()));
        creators.put(DepotViewModel.class,
                () -> new DepotViewModel(App.getDepotRepository()));
        creators.put(CompanyViewModel.class,
                () -> new CompanyViewModel(App.getCompanyRepository()));
        creators.put(OrderViewModel.class,
                () -> new OrderViewModel(App.getTrainRepository()));
        creators.put(AutocompleteViewModel.class,
                () -> new AutocompleteViewModel(App.getTrainRepository()));
        creators.put(AdditionalParamViewModel.class,
                () -> new AdditionalParamViewModel(App.getTempParamRepository()));
        creators.put(TrainInfoViewModel.class,
                () -> new TrainInfoViewModel(App.getTrainRepository()));
        creators.put(KriCoachViewModel.class,
                () -> new KriCoachViewModel(App.getKriCoachRepo()));

        return new CustomViewModelProvider(creators);
    }
}

File: app/src/main/java/com/adrzdv/mtocp/ui/viewmodel/service/ViewModelFactoryProvider.kt
package com.adrzdv.mtocp.ui.viewmodel.service

import androidx.lifecycle.ViewModel
import com.adrzdv.mtocp.AppDependencies
import com.adrzdv.mtocp.ui.viewmodel.model.KriCoachViewModel
import com.adrzdv.mtocp.ui.viewmodel.model.TrainInfoViewModel
import com.adrzdv.mtocp.ui.viewmodel.model.TrainOrderViewModel
import com.adrzdv.mtocp.ui.viewmodel.model.old.AdditionalParamViewModel

class ViewModelFactoryProvider(
    private val appDependencies: AppDependencies
) {
    fun provideFactory(): CustomViewModelProvider {
        val creators: Map<Class<out ViewModel>, Provider<out ViewModel>> = mapOf(
            //ViolationViewModelOld::class.java to Provider { ViolationViewModelOld(appDependencies.violationRepo) },
            //DepotViewModel::class.java to Provider { DepotViewModel(appDependencies.depotRepo) },
            //CompanyViewModel::class.java to Provider { CompanyViewModel(appDependencies.companyRepo) },
            AdditionalParamViewModel::class.java to Provider {
                AdditionalParamViewModel(
                    appDependencies.tempParamRepo
                )
            },
            TrainInfoViewModel::class.java to Provider { TrainInfoViewModel(appDependencies.trainRepo) },
            KriCoachViewModel::class.java to Provider { KriCoachViewModel(appDependencies.kriCoachRepo) },
            TrainOrderViewModel::class.java to Provider {
                TrainOrderViewModel(
                    appDependencies,
                    appDependencies.getDepotByNameUseCase,
                    appDependencies.getTrainByNumberUseCase,
                    appDependencies.createPassengerCoachUseCase,
                    appDependencies.getTrainSchemeUseCase,
                    appDependencies.createDinnerCarUseCase
                )
            }
        )
        return CustomViewModelProvider(creators)
    }

}